File: keyframe.py
"""
* This file is part of PYSLAM 
*
* Copyright (C) 2016-present Luigi Freda <luigi dot freda at gmail dot com> 
*
* PYSLAM is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* PYSLAM is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with PYSLAM. If not, see <http://www.gnu.org/licenses/>.
"""

import cv2
import numpy as np

from scipy.spatial import cKDTree

from threading import RLock

from parameters import Parameters  
from utils import Printer
from collections import defaultdict, OrderedDict, Counter

from frame import Frame 
from camera_pose import CameraPose


class KeyFrameGraph(object):
    def __init__(self):
        self._lock_connections = RLock()           
        # spanning tree
        self.init_parent = False  # is parent initialized? 
        self.parent = None 
        self.children = set()
        # loop edges  
        self.loop_edges = set() 
        self.not_to_erase = False   # if there is a loop edge then you cannot erase this keyframe 
        # covisibility graph 
        self.connected_keyframes_weights = Counter() #defaultdict(int) 
        self.ordered_keyframes_weights = OrderedDict()  # ordered list of connected keyframes (on the basis of the number of map points with this keyframe)
        # 
        self.is_first_connection=True 
        
    # ===============================    
    # spanning tree     
    def add_child(self, keyframe):
        with self._lock_connections:
            self.children.add(keyframe)
    
    def erase_child(self, keyframe):
        with self._lock_connections:
            try: 
                self.children.remove(keyframe)     
            except:
                pass 
            
    def set_parent(self, keyframe):
        with self._lock_connections:
            if self == keyframe: 
                if __debug__:
                    Printer.orange('KeyFrameGraph.set_parent - trying to set self as parent')
                return 
            self.parent = keyframe 
            keyframe.add_child(self)
        
    def get_children(self):
        with self._lock_connections:
            return self.children.copy()
                    
    def get_parent(self):
        with self._lock_connections:
            return self.parent        
        
    def has_child(self, keyframe):            
        with self._lock_connections:
            return keyframe in self.children        
        
               
    # ===============================    
    # loop edges                    
    def add_loop_edge(self, keyframe):
        with self._lock_connections:
            self.not_to_erase = True 
            self.loop_edges.add(keyframe)
        
    def get_loop_edges(self):
        with self._lock_connections:        
            return self.loop_edges.copy()
        
        
    # ===============================    
    # covisibility           
    
    def reset_covisibility(self): 
        self.connected_keyframes_weights = Counter() 
        self.ordered_keyframes_weights = OrderedDict()    
                
    def add_connection(self, keyframe, weigth):
        with self._lock_connections: 
            self.connected_keyframes_weights[keyframe]=weigth
            self.update_best_covisibles()
            
    def erase_connection(self, keyframe):
        with self._lock_connections: 
            try:
                del self.connected_keyframes_weights[keyframe]   
                self.update_best_covisibles()
            except: 
                pass                  
                
    def update_best_covisibles(self):
        with self._lock_connections:         
            self.ordered_keyframes_weights = OrderedDict(sorted(self.connected_keyframes_weights.items(), key=lambda x: x[1], reverse=True)) # order by value (decreasing order)
     
    # get a list of all the keyframe that shares points  
    def get_connected_keyframes(self): 
        with self._lock_connections:          
            return list(self.connected_keyframes_weights.keys()) # returns a copy 
        
    # get an ordered list of covisible keyframes     
    def get_covisible_keyframes(self):
        with self._lock_connections:                  
            return list(self.ordered_keyframes_weights.keys()) # returns a copy   
        
    # get an ordered list of covisible keyframes     
    def get_best_covisible_keyframes(self,N):
        with self._lock_connections:                  
            return list(self.ordered_keyframes_weights.keys())[:N] # returns a copy         
        
    def get_covisible_by_weight(self,weight):  
        with self._lock_connections:          
              return [kf for kf,w in self.ordered_keyframes_weights.items() if w > weight]
          
    def get_weight(self,keyframe): 
        with self._lock_connections:  
            return self.connected_keyframes_weights[keyframe]  
                    

class KeyFrame(Frame,KeyFrameGraph):
    def __init__(self, frame, img=None):
        KeyFrameGraph.__init__(self)
        Frame.__init__(self, img=None, camera=frame.camera, pose=frame.pose, id=frame.id, timestamp=frame.timestamp)   # here we MUST have img=None in order to avoid recomputing keypoint info
                
        if frame.img is not None: 
            self.img = frame.img  # this is already a copy of an image 
        else:
            if img is not None: 
                self.img = img.copy()
                
        self.map = None 
                
        self.is_keyframe = True  
        self.kid = None           # keyframe id 
        
        self._is_bad = False 
        self.to_be_erased = False 
        
        # pose relative to parent (this is computed when bad flag is activated)
        self._pose_Tcp = CameraPose() 

        # share keypoints info with frame (these are computed once for all on frame initialization and they are not changed anymore)
        self.kps     = frame.kps      # keypoint coordinates                  [Nx2]
        self.kpsu    = frame.kpsu     # [u]ndistorted keypoint coordinates    [Nx2]
        self.kpsn    = frame.kpsn     # [n]ormalized keypoint coordinates     [Nx2] (Kinv * [kp,1])    
        self.octaves = frame.octaves  # keypoint octaves                      [Nx1]
        self.sizes   = frame.sizes    # keypoint sizes                        [Nx1] 
        self.angles  = frame.angles   # keypoint angles                       [Nx1] 
        self.des     = frame.des      # keypoint descriptors                  [NxD] where D is the descriptor length 
        
        if hasattr(frame, '_kd'):     
            self._kd = frame._kd 
        else: 
            Printer.orange('KeyFrame %d computing kdtree for input frame %d'%(self.id,frame.id))
            self._kd = cKDTree(self.kpsu)
    
        # map points information arrays (copy points coming from frame)
        self.points   = frame.get_points()     # map points => self.points[idx] is the map point matched with self.kps[idx] (if is not None)
        self.outliers = np.full(self.kpsu.shape[0], False, dtype=bool)     # used just in propagate_map_point_matches()   
        
    # associate matched map points to observations    
    def init_observations(self):
        with self._lock_features:           
            for idx,p in enumerate(self.points):
                if p is not None and not p.is_bad:  
                    if p.add_observation(self, idx):
                        p.update_info()                 
        
    def update_connections(self):
        # for all map points of this keyframe check in which other keyframes they are seen
        # build a counter for these other keyframes    
        points = self.get_matched_good_points()
        assert len(points) > 0
        viewing_keyframes = [kf for p in points for kf in p.keyframes() if kf.kid != self.kid] # exclude this keyframe 
        viewing_keyframes = Counter(viewing_keyframes)   
        if not viewing_keyframes: # if empty   (https://www.pythoncentral.io/how-to-check-if-a-list-tuple-or-dictionary-is-empty-in-python/)
            return 
        # order the keyframes 
        covisible_keyframes = viewing_keyframes.most_common() 
        #print('covisible_keyframes: ', covisible_keyframes)
        # get keyframe that shares most points 
        kf_max, w_max = covisible_keyframes[0]
        # if the counter is greater than threshold add connection
        # otherwise add the one with maximum counter             
        with self._lock_connections:          
            if w_max >= Parameters.kMinNumOfCovisiblePointsForCreatingConnection:
                self.connected_keyframes_weights = viewing_keyframes 
                self.ordered_keyframes_weights = OrderedDict()
                for kf,w in covisible_keyframes:
                    if w >= Parameters.kMinNumOfCovisiblePointsForCreatingConnection:
                        kf.add_connection(self,w)
                        self.ordered_keyframes_weights[kf] = w
                    else:
                        break 
            else:
                self.connected_keyframes_weights = Counter({kf_max,w_max}) 
                self.ordered_keyframes_weights = OrderedDict({kf_max,w_max})    
                kf_max.add_connection(self,w_max)        
            # update spanning tree                     
            if self.is_first_connection and self.kid!=0: 
                self.set_parent(kf_max)
                self.is_first_connection = False 
        #print('ordered_keyframes_weights: ', self.ordered_keyframes_weights)                               
            
    @property             
    def is_bad(self): 
        with self._lock_connections:        
            return self._is_bad        

    def set_not_erase(self): 
        with self._lock_connections:          
            not_to_erase = True 
            
    def set_erase(self): 
        with self._lock_connections: 
            if len(self.loop_edges)==0:          
                self.not_to_erase = False        
            if self.to_be_erased: 
                self.set_bad()        

    def set_bad(self): 
        with self._lock_connections: 
            if self.kid == 0: 
                return 
            if self.not_to_erase: 
                self.to_be_erased = True     
                return
                           
            # update covisibility graph 
            for kf_connected in list(self.connected_keyframes_weights.keys()):          
                kf_connected.erase_connection(self)    
                
            for idx,p in enumerate(self.points): 
                if p is not None: 
                    p.remove_observation(self,idx)
                                 
            self.reset_covisibility()
            
            # update spanning tree: each children must be connected to a new parent 
            
            # build a set of parent candidates for the children 
            parent_candidates = set() 
            assert(self.parent is not None)
            parent_candidates.add(self.parent)
            
            # each child must be connected to a new parent (the candidate parent with highest covisibility weight)
            # once a child is connected to a new parent, include the child as new parent candidate for the rest            
            while not len(self.children)==0:            
                w_max = 0
                child_to_connect = None 
                parent_to_connect = None 
                found_connection = False 
                for kf_child in self.children: 
                    if kf_child.is_bad:
                        continue
                    # check if a candidate parent is connected to kf_child and compute the candidate parent with highest covisibility weight                     
                    covisible_keyframes = kf_child.get_covisible_keyframes()
                    for candidate_parent in parent_candidates: 
                        if candidate_parent in covisible_keyframes:
                            w = kf_child.get_weight(candidate_parent)
                            if w > w_max: 
                                w_max = w 
                                child_to_connect = kf_child
                                parent_to_connect = candidate_parent 
                                found_connection = True 
                if found_connection: 
                    child_to_connect.set_parent(parent_to_connect)
                    parent_candidates.add(child_to_connect)
                    self.children.remove(child_to_connect)
                else: 
                    break # stop since there is no connection with covisibility weight>0

            # if a child has no covisibility connections with any parent candidate, connect it with the original parent of this keyframe
            if not len(self.children)==0:
                for kf_child in self.children:  
                    kf_child.set_parent(self.parent)
                    
            self.parent.erase_child(self)
            self._pose_Tcp.update(self.Tcw @ self.parent.Twc)
            self._is_bad = True 
            
        if self.map is not None:
            self.map.remove_keyframe(self)

File: pyslam/slam/cpp/keyframe.cpp
/*
 * This file is part of PYSLAM
 *
 * Copyright (C) 2016-present Luigi Freda <luigi dot freda at gmail dot com>
 *
 * PYSLAM is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * PYSLAM is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with PYSLAM. If not, see <http://www.gnu.org/licenses/>.
 */

#include "keyframe.h"
#include "camera_pose.h"
#include "config_parameters.h"
#include "frame.h"
#include "map.h"
#include "map_point.h"
#include "utils/messages.h"

#include <nlohmann/json.hpp>

#include <atomic>
#include <mutex>

namespace pyslam {

// Static member definitions
std::atomic<int> KeyFrame::next_kid_{0};
std::mutex KeyFrame::kid_mutex_;

// ======================================================
// KeyFrameGraph implementation
// ======================================================

KeyFrameGraph::KeyFrameGraph() : parent(nullptr) {}

void KeyFrameGraph::add_child(KeyFramePtr child) {
    std::lock_guard<std::mutex> lock(_lock_connections);
    add_child_no_lock(child);
}

void KeyFrameGraph::add_child_no_lock(KeyFramePtr child) {
    if (std::find(children.begin(), children.end(), child) == children.end()) {
        children.insert(child);
    }
}

void KeyFrameGraph::erase_child(KeyFramePtr child) {
    std::lock_guard<std::mutex> lock(_lock_connections);
    erase_child_no_lock(child);
}

void KeyFrameGraph::erase_child_no_lock(KeyFramePtr child) {
    auto it = std::find(children.begin(), children.end(), child);
    if (it != children.end()) {
        children.erase(it);
    }
}

void KeyFrameGraph::set_parent_no_lock(KeyFramePtr parent) {

    if (parent.get() == this) {
        return;
    }
    // Set the parent
    this->parent = parent;
    auto this_as_kf = std::dynamic_pointer_cast<KeyFrame>(this->shared_from_this());
    if (this_as_kf == nullptr) {
        throw std::runtime_error("KeyFrameGraph is not a KeyFrame");
    }
    auto self = this_as_kf->KeyFrameGraph::downcasted_shared_from_this<KeyFrame>();
    if (!self) {
        MSG_ERROR("KeyFrameGraph could not be downcasted to KeyFrame");
        return;
    }
    // Add this keyframe as a child of the parent
    parent->add_child(self);
}

void KeyFrameGraph::set_parent(KeyFramePtr parent) {
    std::lock_guard<std::mutex> lock(_lock_connections);
    set_parent_no_lock(parent);
}

std::set<KeyFramePtr> KeyFrameGraph::get_children() const {
    std::lock_guard<std::mutex> lock(_lock_connections);
    return children;
}

KeyFramePtr KeyFrameGraph::get_parent() const {
    std::lock_guard<std::mutex> lock(_lock_connections);
    return parent;
}

bool KeyFrameGraph::has_child(const KeyFramePtr keyframe) const {
    std::lock_guard<std::mutex> lock(_lock_connections);
    return std::find(children.begin(), children.end(), keyframe) != children.end();
}

void KeyFrameGraph::add_loop_edge(KeyFramePtr kf) {
    std::lock_guard<std::mutex> lock(_lock_connections);
    not_to_erase = true;
    // Add the loop edge only if it is not already in the set
    if (std::find(loop_edges.begin(), loop_edges.end(), kf) == loop_edges.end()) {
        loop_edges.insert(kf);
    }
}

std::set<KeyFramePtr> KeyFrameGraph::get_loop_edges() const {
    std::lock_guard<std::mutex> lock(_lock_connections);
    return loop_edges;
}

void KeyFrameGraph::add_connection_no_lock(KeyFramePtr kf, int weight) {
    connected_keyframes_weights[kf] = weight;
    update_best_covisibles_no_lock_();
}

void KeyFrameGraph::add_connection(KeyFramePtr kf, int weight) {

    std::lock_guard<std::mutex> lock(_lock_connections);
    add_connection_no_lock(kf, weight);
}

void KeyFrameGraph::update_best_covisibles_no_lock_() {

    // Convert map to vector for sorting
    std::vector<std::pair<KeyFramePtr, int>> covisibles_vec;
    covisibles_vec.reserve(connected_keyframes_weights.size());
    for (const auto &pair : connected_keyframes_weights) {
        covisibles_vec.push_back(pair);
    }

    // Sort the vector
    std::sort(covisibles_vec.begin(), covisibles_vec.end(),
              [](const std::pair<KeyFrameGraphPtr, int> &a,
                 const std::pair<KeyFrameGraphPtr, int> &b) { return a.second > b.second; });

    // Update ordered_keyframes_weights
    ordered_keyframes_weights = std::move(covisibles_vec);
}

// KeyFrameGraph missing methods
void KeyFrameGraph::reset_covisibility() {
    connected_keyframes_weights.clear();
    ordered_keyframes_weights.clear();
}

void KeyFrameGraph::erase_connection_no_lock(KeyFramePtr keyframe) {
    auto it = connected_keyframes_weights.find(keyframe);
    if (it != connected_keyframes_weights.end()) {
        connected_keyframes_weights.erase(it);
        update_best_covisibles_no_lock_();
    }
}

void KeyFrameGraph::erase_connection(KeyFramePtr keyframe) {
    std::lock_guard<std::mutex> lock(_lock_connections);
    erase_connection_no_lock(keyframe);
}

std::vector<KeyFramePtr> KeyFrameGraph::get_connected_keyframes_no_lock() const {
    std::vector<KeyFramePtr> connected;
    for (const auto &pair : connected_keyframes_weights) {
        connected.push_back(pair.first);
    }
    return connected;
}

std::vector<KeyFramePtr> KeyFrameGraph::get_connected_keyframes() const {
    std::lock_guard<std::mutex> lock(_lock_connections);
    return get_connected_keyframes_no_lock();
}

std::vector<KeyFramePtr> KeyFrameGraph::get_covisible_keyframes_no_lock() const {
    std::vector<KeyFramePtr> covisible;
    covisible.reserve(ordered_keyframes_weights.size());
    for (const auto &pair : ordered_keyframes_weights) {
        covisible.push_back(pair.first);
    }
    return covisible;
}

std::vector<KeyFramePtr> KeyFrameGraph::get_covisible_keyframes() const {
    std::lock_guard<std::mutex> lock(_lock_connections);
    return get_covisible_keyframes_no_lock();
}

std::vector<KeyFramePtr> KeyFrameGraph::get_best_covisible_keyframes(int N) const {
    std::lock_guard<std::mutex> lock(_lock_connections);
    std::vector<KeyFramePtr> best_covisible;
    best_covisible.reserve(std::min(N, static_cast<int>(ordered_keyframes_weights.size())));
    int count = 0;
    for (const auto &pair : ordered_keyframes_weights) {
        if (count >= N)
            break;
        best_covisible.push_back(pair.first);
        count++;
    }
    return best_covisible;
}

std::vector<KeyFramePtr> KeyFrameGraph::get_covisible_by_weight(int weight) const {
    std::lock_guard<std::mutex> lock(_lock_connections);
    std::vector<KeyFramePtr> covisible;
    for (const auto &pair : ordered_keyframes_weights) {
        if (pair.second > weight) {
            covisible.push_back(pair.first);
        }
    }
    return covisible;
}

int KeyFrameGraph::get_weight_no_lock(const KeyFramePtr &keyframe) const {
    auto it = connected_keyframes_weights.find(keyframe);
    return (it != connected_keyframes_weights.end()) ? it->second : 0;
}

int KeyFrameGraph::get_weight(const KeyFramePtr &keyframe) const {
    std::lock_guard<std::mutex> lock(_lock_connections);
    return get_weight_no_lock(keyframe);
}

std::unordered_map<int, int> KeyFrameGraph::get_connected_keyframes_weights() const {
    std::lock_guard<std::mutex> lock(_lock_connections);
    std::unordered_map<int, int> weights_map;
    for (const auto &[kf, weight] : connected_keyframes_weights) {
        if (kf) {
            weights_map[kf->id] = weight;
        }
    }
    return weights_map;
}

// ======================================================
// KeyFrame implementation
// ======================================================

// KeyFrame implementation
KeyFrame::KeyFrame(const FramePtr &frame, const cv::Mat &img, const cv::Mat &img_right,
                   const cv::Mat &depth, int kid)
    : Frame(frame->camera, cv::Mat(), cv::Mat(), cv::Mat(), CameraPose(), frame ? frame->id : -1,
            frame ? frame->timestamp : 0.0, frame ? frame->img_id : -1),
      kid(kid), _is_bad(false), to_be_erased(false), lba_count(0), loop_query_id(-1),
      num_loop_words(0), loop_score(0.0f), reloc_query_id(-1), num_reloc_words(0),
      reloc_score(0.0f), GBA_kf_id(0), map(nullptr) {

    if (frame) {
        Frame::copy_from(*frame);
    }
    // We don't preserve KeyFrame outliers and reinitialize all the flags to false
    outliers = std::vector<bool>(frame->outliers.size(), false);

    // Set as keyframe
    is_keyframe = true;

    // Initialize pose relative to parent - use default constructor, not
    // make_unique
    _pose_Tcp = CameraPose();

    // Initialize GBA poses - use default constructor, not make_unique
    is_Tcw_GBA_valid = false;
    Tcw_GBA = Eigen::Matrix4d::Identity();
    Tcw_before_GBA = Eigen::Matrix4d::Identity();

    // Copy images if provided
    if (!img.empty()) {
        this->img = img.clone();
    }
    if (!img_right.empty()) {
        this->img_right = img_right.clone();
    }
    if (!depth.empty()) {
        this->depth_img = depth.clone();
    }
}

void KeyFrame::init_observations() {
    std::lock_guard<std::mutex> lock(_lock_features);
    auto self = KeyFrameGraph::downcasted_shared_from_this<KeyFrame>();
    if (!self) {
        MSG_ERROR("KeyFrameGraph could not be downcasted to KeyFrame");
        return;
    }
    for (size_t idx = 0; idx < points.size(); ++idx) {
        const auto &p = points[idx];
        if (p && !p->is_bad()) {
            if (p->add_observation(self, static_cast<int>(idx))) {
                p->update_info();
            }
        }
    }
}

void KeyFrame::update_connections() {
    // for all map points of this keyframe check in which other keyframes they are seen
    // build a counter for these other keyframes

    // Get all matched good points from this keyframe
    const auto points = get_matched_good_points();
    int num_points = static_cast<int>(points.size());

    if (num_points == 0) {
        // Equivalent to Printer.orange("KeyFrame: update_connections - frame without points")
        MSG_WARN("KeyFrame: update_connections - frame without points");
        return;
    }

    // Build a counter for viewing keyframes with stable (first-seen) order
    std::unordered_map<KeyFramePtr, int> viewing_keyframes;
    for (const auto &p : points) {
        const auto point_keyframes = p->keyframes();
        for (const auto &kf : point_keyframes) {
            if (kf->kid != this->kid) {
                viewing_keyframes[kf]++;
            }
        }
    }

    if (viewing_keyframes.empty()) {
        return;
    }

    // Order the keyframes: sort by weight in descending order
    // Convert to vector for sorting (equivalent to most_common() in Python Counter)
    std::vector<std::pair<KeyFramePtr, int>> covisible_keyframes;
    covisible_keyframes.reserve(viewing_keyframes.size());
    for (const auto &pair : viewing_keyframes) {
        covisible_keyframes.push_back(pair);
    }

    // Sort the vector by weight in descending order
    std::sort(covisible_keyframes.begin(), covisible_keyframes.end(),
              [](const auto &a, const auto &b) { return a.second > b.second; }); // descending order

    auto kf_max = covisible_keyframes[0].first;
    auto w_max = covisible_keyframes[0].second;

    auto self = KeyFrameGraph::downcasted_shared_from_this<KeyFrame>();
    if (!self) {
        MSG_ERROR("KeyFrameGraph could not be downcasted to KeyFrame");
        return;
    }

    std::lock_guard<std::mutex> lock(_lock_connections);

    connected_keyframes_weights = std::move(viewing_keyframes);

    if (w_max >= Parameters::kMinNumOfCovisiblePointsForCreatingConnection) {
        ordered_keyframes_weights.clear();
        ordered_keyframes_weights.reserve(covisible_keyframes.size());

        // Here we keep the weight-decreasing-order of the covisible_keyframes
        for (const auto &[kf, w] : covisible_keyframes) {
            if (w >= Parameters::kMinNumOfCovisiblePointsForCreatingConnection) {
                kf->add_connection_no_lock(self, w);
                ordered_keyframes_weights.push_back({kf, w});
            } else {
                break; // Since sorted, no more will meet threshold
            }
        }
    } else {
        ordered_keyframes_weights.clear();
        ordered_keyframes_weights.push_back({kf_max, w_max});

        kf_max->add_connection_no_lock(self, w_max);
    }

    // Update spanning tree
    // We need to avoid setting the parent to None or self or a bad keyframe
    if (is_first_connection && (kid != 0) && (kf_max) && ((kf_max->id != this->id)) &&
        !kf_max->is_bad()) {
        set_parent_no_lock(kf_max);
        is_first_connection = false;
    }
}

bool KeyFrame::is_bad() const {
    std::lock_guard<std::mutex> lock(_lock_connections);
    return _is_bad;
}

Eigen::Matrix4d KeyFrame::Tcp() const {
    std::lock_guard<std::mutex> lock(_lock_connections);
    return _pose_Tcp.get_matrix();
}

void KeyFrame::set_not_erase() {
    std::lock_guard<std::mutex> lock(_lock_connections);
    not_to_erase = true;
}

void KeyFrame::set_erase() {
    {
        std::lock_guard<std::mutex> lock(_lock_connections);
        if (loop_edges.empty()) {
            not_to_erase = false;
        }
    }
    if (to_be_erased) {
        set_bad();
    }
}

void KeyFrame::set_bad() {
    std::lock_guard<std::mutex> lock(_lock_connections);
    if (kid <= 0) {
        return;
    }

    if (not_to_erase) {
        to_be_erased = true;
        return;
    }
    auto self = KeyFrameGraph::downcasted_shared_from_this<KeyFrame>();
    if (!self) {
        MSG_ERROR("KeyFrameGraph could not be downcasted to KeyFrame");
        return;
    }

    // 1) Remove covisibility connections
    auto connected_keyframes = get_connected_keyframes_no_lock();
    for (auto &kf_connected : connected_keyframes) {
        kf_connected->erase_connection_no_lock(self);
    }

    // 2) Remove feature observations
    for (size_t idx = 0; idx < points.size(); ++idx) {
        const auto &p = points[idx];
        if (p) {
            p->remove_observation(self, static_cast<int>(idx));
        }
    }

    reset_covisibility();

    // 3) Update spanning tree
    MSG_FORCED_ASSERT(parent, "KeyFrame: set_bad - parent is nullptr");

    std::set<KeyFramePtr> parent_candidates = {std::static_pointer_cast<KeyFrame>(parent)};

    std::vector<KeyFramePtr> remaining_children(children.begin(), children.end());
    children.clear();

    // Reassign children via covisibility to any parent candidate; one child per iteration
    int iters = 0;
    const int max_iters = static_cast<int>(remaining_children.size()) * 100;
    while (!remaining_children.empty() && iters < max_iters) {
        iters++;

        KeyFramePtr best_child;
        KeyFramePtr best_parent;
        int max_weight = -1;

        // Find best reassignment across all remaining children
        for (auto &child : remaining_children) {
            if (child->is_bad())
                continue;

            auto covisibles = child->get_covisible_keyframes_no_lock();
            for (auto &candidate : parent_candidates) {
                if (std::find(covisibles.begin(), covisibles.end(), candidate) !=
                    covisibles.end()) {
                    int w = child->get_weight_no_lock(candidate);
                    if (w > max_weight) {
                        max_weight = w;
                        best_child = child;
                        best_parent = candidate;
                    }
                }
            }
        }

        // If no covisible parent candidate found, stop trying to reassign via covisibility
        if (!best_child || !best_parent)
            break;

        // Commit the best reassignment found in this iteration
        best_child->set_parent(best_parent);
        parent_candidates.insert(best_child);

        // Safely remove the chosen child (no range-iteration here)
        auto it = std::find(remaining_children.begin(), remaining_children.end(), best_child);
        if (it != remaining_children.end())
            remaining_children.erase(it);

        if (iters >= max_iters) {
            MSG_WARN("KeyFrame: set_bad - max iterations reached");
            break;
        }
    }

    // 4) Reassign any still-unconnected children to the original parent
    for (auto &child : remaining_children) {
        child->set_parent_no_lock(parent);
    }

    // 5) Cleanup
    auto parent_kf = std::static_pointer_cast<KeyFrame>(parent);
    parent->erase_child_no_lock(self);
    _pose_Tcp.update(this->Tcw() * parent_kf->Twc());
    _is_bad = true;

    if (map) {
        map->remove_keyframe(self);
    }
}

} // namespace pyslam
